data-plane/receiver-loom/src/main/java/dev/knative/eventing/kafka/broker/receiverloom/LoomKafkaProducer.java

```diff
@@ -15,6 +15,132 @@
  */
 package dev.knative.eventing.kafka.broker.receiverloom;
 
-public class LoomKafkaProducer<K, V> {
-    // TODO implement
+import dev.knative.eventing.kafka.broker.core.ReactiveKafkaProducer;
+import dev.knative.eventing.kafka.broker.core.tracing.kafka.ProducerTracer;
+import io.vertx.core.Future;
+import io.vertx.core.Promise;
+import io.vertx.core.Vertx;
+import io.vertx.core.impl.ContextInternal;
+import io.vertx.core.impl.VertxInternal;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.apache.kafka.clients.producer.Producer;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.clients.producer.RecordMetadata;
+
+public class LoomKafkaProducer<K, V> implements ReactiveKafkaProducer<K, V> {
+
+    private final Producer<K, V> producer;
+
+    private final BlockingQueue<RecordPromise> eventQueue;
+    private final AtomicBoolean isClosed;
+    private final ProducerTracer tracer;
+    private final VertxInternal vertx;
+    private final ContextInternal ctx;
+
+    public LoomKafkaProducer(Vertx v, Producer<K, V> producer) {
+        this.producer = producer;
+        this.eventQueue = new LinkedBlockingQueue<>();
+        this.isClosed = new AtomicBoolean(false);
+        this.vertx = (VertxInternal) v;
+
+        if (v != null) {
+            ContextInternal ctxInt = ((ContextInternal) v.getOrCreateContext()).unwrap();
+            this.tracer = ProducerTracer.create(ctxInt.tracer());
+            this.ctx = vertx.getOrCreateContext();
+        } else {
+            this.tracer = null;
+            this.ctx = null;
+        }
+        Thread.ofVirtual().start(this::sendFromQueue);
+    }
+
+    @Override
+    public Future<RecordMetadata> send(ProducerRecord<K, V> record) {
+        Promise<RecordMetadata> promise = Promise.promise();
+        if (isClosed.get()) {
+            promise.fail("Producer is closed");
+        } else {
+            eventQueue.add(new RecordPromise(record, promise));
+        }
+        return promise.future();
+    }
+
+    private void sendFromQueue() {
+        while (true) {
```

Comment:
Did after close() sendFromQueue VThread will get end or it will get preempted in the background, May need to change the condition.

Reply:
In addition, we need to interrupt the thread on close because it might be waiting on `take()` and it will never exit

Reply:
> In addition, we need to interrupt the thread on close because it might be waiting on take() and it will never exit

Oh I see, it might not work when the producer is not closed and we are waiting on queue.take() and without any incoming event we are closing.